### ISSUE TITLE
A complete example of a dummy Array wrapper type, with broadcasting, reductions, etc.

### ISSUE THREAD

**Issue by kcajf** [JuliaLang/julia#33453]
I think it would be very useful to have, somewhere in the docs, a complete example of a custom array type (trivially, a wrapper around an `Array` with a token amount of extra metadata) that does the following:

- correct, fast `getindex` and `setindex!` (correct use of `@propagate_inbounds`, `@boundscheck`, etc)
- correct implementation of different `similar` methods
- fully implements broadcasting, such that adding two of these custom wrapper types together will add together the wrapped and place that in the appropriate wrapper type. Should also demonstrate how to make your custom type broadcast correctly with any other AbstractArray.
- fully implements the necessary machinery for generic reduction operations to work and preserve the wrapper type. I.e.  `sum(x, dims=2)` on a 2d custom array should construct and allocate into an appropriately sized 1d version of the wrapper type.

This information is currently spread in a few places in the docs, but is often quite vague. In order to work all this stuff out, one has to mainly rely on reading other people's custom array types (some of which has mistakes / is out of date / incomplete), or reading source code in base.

**Comment by mbauman** [comment]
This is tough to do because everything other than size and get/setindex is optional and there are lots of dimensions of variation.  Exactly which methods beyond that you want to specialize is going to be different for each and every use-case — even simple wrappers.

To be concrete with one of the bullets you've listed — once you've defined the above three methods, broadcasting will _just work_.  How you want to hook into broadcast will vary based upon how you want your array to behave.  You could act like some other array, or opt-out of fusion, or you could just have a custom return type, or you could have a wholly custom implementation.  You could choose to interact with other arrays in the broadcast or not.  The goal is to have this all described in a reasonable manner — with examples — in the [Broadcasting section of the Interfaces chapter](https://docs.julialang.org/en/v1/manual/interfaces/#man-interfaces-broadcasting-1), but it is really hard.  Improvements here are welcome.

In other words, what you consider "complete" will be very different from what someone else wants. There's a very long tail of `AbstractArray` functionality — and it can all be piecemeal specialized.  Doing the work to e.g., specify and document an interface to provide a whole cluster of functionality at once (like reductions) is itself a large task.  That's not to say it shouldn't be done!  But rather the opposite — I'd want to see reductions appear as a chapter in Interfaces instead of just giving an ad-hoc example mixed in with other things.

We could add links to registered array packages that do particular things well as examples.

**Comment by vpuri3** [comment]
check out https://github.com/SciML/RecursiveArrayTools.jl/blob/master/src/array_partition.jl for a very complete array wrapper.

also check out https://github.com/vpuri3/SpectralElements.jl/blob/vp-updates/src/Spectral/Field.jl . This is WIP
